syft/execution/protocol.py

```diff
@@ -1,479 +1,560 @@
+        #         warnings.warn(f"Failed to translate Protocol with {translator}")
+
+        return results
+
+    def toggle_tracing(self, value=None):
+        self.tracing = value if value is not None else not self.tracing
+        # self.state.tracing = self.tracing
+        for role in self.roles.values():
+            for ph in role.placeholders.values():
+                ph.tracing = self.tracing
+
+    def copy(self):
+        """Creates a copy of a protocol."""
+        protocol_copy = Protocol(
+            name=self.name,
+            roles={role_id: role.copy() for role_id, role in self.roles.items()},
+            include_state=self.include_state,
+            is_built=self.is_built,
+            id=sy.ID_PROVIDER.pop(),
+            owner=self.owner,
+            tags=self.tags,
+            description=self.description,
+        )
 
-    def run(self, *args, **kwargs):
+        protocol_copy.torchscript = self.torchscript
+
+        return protocol_copy
+
+    # def __setattr__(self, name, value):
+    #     """Add new tensors or parameter attributes to the state and register them
+    #     in the owner's registry
+    #     """
+    #     if isinstance(value, torch.jit.ScriptModule):
+    #         object.__setattr__(self, name, value)
+    #     elif isinstance(value, FrameworkTensor):
+    #         self.role.register_state_tensor(value, self.owner)
+    #         self.state_attributes[name] = value
+    #     elif isinstance(value, FrameworkLayerModule):
+    #         for param in value.parameters():
+    #             self.role.register_state_tensor(param, self.owner)
+    #         self.state_attributes[name] = value
+    #     else:
+    #         object.__setattr__(self, name, value)
+
+    # def __getattr__(self, name):
+    #     if name not in self.state_attributes:
+    #         raise AttributeError("State attribute not found.")
+
+    #     value = self.state_attributes[name]
+    #     if not self.is_building:
+    #         return value
+
+    #     if isinstance(value, FrameworkTensor):
+    #         return self.role.placeholders[value.id]
+    #     elif isinstance(value, FrameworkLayerModule):
+    #         # We need to deepcopy here otherwise the real layer is modified when the Protocol is being built
+    #         copied_layer = copy.deepcopy(value)
+    #         for copied_param, param in zip(copied_layer.named_parameters(), value.parameters()):
+    #             (copied_name, _) = copied_param
+    #             copied_layer._parameters[copied_name] = self.role.placeholders[param.id]
+
+    #         return copied_layer
+
+    def __call__(self, *args):
         """
-        Run the protocol by executing the plans sequentially
+        Calls a protocol execution with some arguments.
+
+        When possible, run the original function to improve efficiency. When
+        it's not, for example if you fetched the protocol from a remote worker,
+        then run it from the tape of actions:
```

Comment:
Maybe something like?

```suggestion
        Run actions on the workers provided for each Role from the Role's tape of actions
```

Reply:
Thanks! And sorry for being responsible of half the obsolete docstrings in the codebase 🙁